src/java.base/share/classes/jdk/internal/misc/Blocker.java

```diff
@@ -0,0 +1,143 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.misc;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.security.AccessController;
+import java.security.PrivilegedExceptionAction;
+import java.util.concurrent.ForkJoinPool;
+import jdk.internal.access.JavaLangAccess;
+import jdk.internal.access.SharedSecrets;
+
+/**
+ * Defines static methods to mark the beginning and end of a possibly blocking
+ * operation. The methods are intended to be used with try-finally as follows:
+ * {@snippet lang=java :
+ *     long comp = Blocker.begin();
+ *     try {
+ *         // blocking operation
+ *     } finally {
+ *         Blocker.end(comp);
+ *     }
+ * }
+ * If invoked from a virtual thread and the underlying carrier thread is a
+ * CarrierThread then the code in the block runs as if it were in run in
+ * ForkJoinPool.ManagedBlocker. This means the pool can be expanded to support
+ * additional parallelism during the blocking operation.
+ */
+public class Blocker {
+    private static final JavaLangAccess JLA;
+    static {
+        JLA = SharedSecrets.getJavaLangAccess();
+        if (JLA == null) {
+            throw new InternalError("JavaLangAccess not setup");
+        }
+    }
+
+    private Blocker() { }
+
+    private static Thread currentCarrierThread() {
+        return JLA.currentCarrierThread();
+    }
+
+    /**
+     * Marks the beginning of possibly blocking operation.
+     * @return the return value from the attempt to compensate
+     */
+    public static long begin() {
+        if (VM.isBooted()
+                && currentCarrierThread() instanceof CarrierThread ct && !ct.inBlocking()) {
+            ct.beginBlocking();
+            long comp = ForkJoinPools.beginCompensatedBlock(ct.getPool());
```

Comment:
It appears that `ForkJoinPools.beginCompensatedBlock(...)` can throw an exception. Would such an exception require the CarrierThread's blocking state (which we set on the previous line) to be reset by a call to `ct.endBlocking()`? Or is it futile to reset that state if any exception gets thrown from the call to `ForkJoinPools.beginCompensatedBlock(...)`?

Reply:
REE isn't possible here but maybe you mean a resource issues such as stack overflow or OOME. If that happens then the flag wouldn't be reset. It wouldn't a corrects issue but we may be able to make this a bit more robust for these cases.

Reply:
Hello Alan, 
My concern wasn't really about OOME or resource issue. I had noticed the `ForkJoinPools.beginCompensatedBlock` has a try/catch which then propagates back the `Throwable`. So I thought there could be some genuine exceptions that could be thrown in this flow which ultimately calls `ForkJoinPool.beginCompensatedBlock()`. I haven't yet reached the changes in `ForkJoinPool.java` class itself, so it's likely that this catch block is only there for theoretical reasons when dealing with `MethodHandle`.